models/classes/pack/ItemPack.php

```diff
@@ -128,40 +130,50 @@ public function getData()
      * Set item's assets of a given type to the pack.
      *
      * @param string $type the assets type, one of those who are supported.
-     * @param string[] $assets the list of assets' URL to load
-     *
-     * @param \tao_models_classes_service_StorageDirectory $publicDirectory
+     * @param array $assets the list of assets' URL to load
      *
      * @throw InvalidArgumentException
      */
-    public function setAssets($type, $assets, $publicDirectory)
+    public function setAssets(string $type, $assets): void
     {
-        if (!in_array($type, self::$assetTypes)) {
-            throw new InvalidArgumentException('Unknow asset type "' . $type . '", it should be either ' . implode(', ', self::$assetTypes));
-        }
         if (!is_array($assets)) {
             throw new InvalidArgumentException('Assests should be an array, "' . gettype($assets) . '" given');
         }
 
-        /**
-         * Apply active encoder immediately
-         * @var Encoding $encoder
-         */
-        $encoder = EncoderService::singleton()->get($this->assetEncoders[$type], $publicDirectory);
         foreach ($assets as $asset) {
-            if ($asset instanceof MediaAsset) {
-                $mediaSource = $asset->getMediaSource();
-                if ($mediaSource instanceof MediaSource || $mediaSource instanceof HttpSource) {
-                    $assetKey = $asset->getMediaIdentifier();
-                } else {
-                    $assetKey = $mediaSource->getBaseName($asset->getMediaIdentifier());
-                }
+            $this->setAsset($type, $asset);
+        }
+    }
+
+    /**
+     * @param string $type
+     * @param string|MediaAsset $asset
+     *
+     * @throw InvalidArgumentException
+     */
+    public function setAsset(string $type, $asset): void
+    {
+        if (!in_array($type, self::$assetTypes)) {
+            throw new InvalidArgumentException('Unknow asset type "' . $type . '", it should be either ' . implode(', ', self::$assetTypes));
+        }
+
+        /** @var Encoding $encoder */
+        $encoder = EncoderService::singleton()->get($this->assetEncoders[$type]);
+
+        if ($asset instanceof MediaAsset) {
```

Comment:
I agree with @bartlomiejmarszal. We could make a method just to discover assetKey. This we we reduce method complexity (nested if/else) and is is easier to read. A suggestion:

```
    private function getAssetKey($asset): string
    {
        if (!$asset instanceof MediaAsset) {
            return $asset;
        }

        $mediaSource = $asset->getMediaSource();

        if ($mediaSource instanceof MediaSource || $mediaSource instanceof HttpSource) {
            return $asset->getMediaIdentifier();
        }

        return $mediaSource->getBaseName($asset->getMediaIdentifier());
    }
```

Reply:
no prob, it is good suggestion